File: todo-api-js/src/application/repositories/todo.ts
import { Exception, type InternalError } from "@domain/utils/exception";
import { type Result } from "@domain/utils/result";
import { type Id } from "@domain/value-objects/id";
import { type Status, type Description, type Title, type TodoEntity } from "@domain/entities/todo";

export interface TodoRepository {
    findById(todoId: Id): Promise<Result<TodoEntity, FindByIdError>>;
    create(entity: TodoEntity): Promise<Result<void, CreateError>>;
    update(entity: UpdateQuery): Promise<Result<void, UpdateError>>;
    deleteById(todoId: Id): Promise<Result<void, DeleteByIdError>>;
    listPaginated(query: ListPaginatedQuery): Promise<Result<PaginatedList, ListError>>;
}

export type UpdateQuery = {
    id: Id;
    title: Title;
    description?: Description;
    status: Status;
    todoAt?: Date;
};

export type ListPaginatedQuery = { page: number; limit: number };

// TODO: maybe add a common type to this
export type PaginatedList = {
    count: number;
    page: number;
    limit: number;
    data: TodoEntity[];
};

export type FindByIdError = IdNotFound | InternalError;

export type CreateError = DuplicatedTitle | InternalError;

export type UpdateError = DuplicatedTitle | IdNotFound | InternalError;

export type DeleteByIdError = IdNotFound | InternalError;

export type ListError = InternalError;

export class DuplicatedTitle extends Exception<"DuplicatedTitle"> {
    constructor(title: Title) {
        super("DuplicatedTitle", `Todo with title ${title.value} already exists`);
    }
}

export class IdNotFound extends Exception<"IdNotFound"> {
    constructor(id: Id) {
        super("IdNotFound", `Todo with id ${id.value} could not be found`);
    }
}

File: todo-api-rs/src/application/repositories/todo.rs
use std::error;
use std::num::NonZeroU32;

use thiserror::Error;

use crate::domain::entities::todo::{Description, Title, TodoEntity, Status};
use crate::domain::types::{Date, Id};

pub trait TodoRepository {
    async fn create(&mut self, todo: TodoEntity) -> Result<(), CreateError>;
    async fn delete(&mut self, todo_id: Id) -> Result<(), DeleteError>;
    async fn find(&self, todo_id: Id) -> Result<TodoEntity, FindError>;
    async fn list(&self, query: ListQuery) -> Result<PaginatedList, ListError>;
    async fn update(&mut self, query: UpdateQuery) -> Result<(), UpdateError>;
}

#[derive(Clone, Debug)]
pub struct UpdateQuery {
    pub id: Id,
    pub title: Title,
    pub description: Option<Description>,
    pub status: Status,
    pub todo_at: Option<Date>,
}

#[derive(Clone, Debug)]
pub struct ListQuery {
    pub page: NonZeroU32,
    pub per_page: NonZeroU32,
    pub title: Option<Title>,
}

#[derive(Clone, Debug)]
pub struct PaginatedList {
    pub count: u64,
    pub items: Vec<TodoEntity>,
}

#[derive(Debug, Error)]
pub enum CreateError {
    #[error("Todo title already exists")]
    DuplicatedTitle,
    #[error(transparent)]
    Internal(Box<dyn error::Error>),
}

#[derive(Debug, Error)]
pub enum DeleteError {
    #[error("Todo could not be found")]
    NotFound,
    #[error(transparent)]
    Internal(Box<dyn error::Error>),
}

#[derive(Debug, Error)]
pub enum FindError {
    #[error("Todo could not be found")]
    NotFound,
    #[error(transparent)]
    Internal(Box<dyn error::Error>),
}

#[derive(Debug, Error)]
pub enum ListError {
    #[error(transparent)]
    Internal(Box<dyn error::Error>),
}

#[derive(Debug, Error)]
pub enum UpdateError {
    #[error("Todo could not be found")]
    NotFound,
    #[error("Todo title already exists")]
    DuplicatedTitle,
    #[error(transparent)]
    Internal(Box<dyn error::Error>),
}
